src/duct.py

```diff
@@ -310,53 +314,58 @@ class TailPipe:
 
     TAIL_CYCLE_TIME = 0.01
 
-    def __init__(self, file_path, buffer):
+    def __init__(self, file_path: str, buffer: IO[bytes]) -> None:
         self.file_path = file_path
         self.buffer = buffer
-        self.stop_event = None
-        self.infile = None
-        self.thread = None
+        self.stop_event: threading.Event | None = None
+        self.infile: IO[bytes] | None = None
+        self.thread: threading.Thread | None = None
 
-    def start(self):
+    def start(self) -> None:
         Path(self.file_path).touch()
         self.stop_event = threading.Event()
         self.infile = open(self.file_path, "rb")
         self.thread = threading.Thread(target=self._tail, daemon=True)
         self.thread.start()
 
-    def fileno(self):
+    def fileno(self) -> int:
+        assert self.infile is not None
         return self.infile.fileno()
 
-    def _catch_up(self):
+    def _catch_up(self) -> None:
+        assert self.infile is not None
         data = self.infile.read()
         if data:
             self.buffer.write(data)
             self.buffer.flush()
 
-    def _tail(self):
+    def _tail(self) -> None:
+        assert self.stop_event is not None
         try:
             while not self.stop_event.is_set():
                 self._catch_up()
                 time.sleep(TailPipe.TAIL_CYCLE_TIME)
-
             # After stop event, collect and passthrough data one last time
             self._catch_up()
         except Exception:
             raise
         finally:
             self.buffer.flush()
 
-    def close(self):
+    def close(self) -> None:
+        assert self.stop_event is not None
+        assert self.thread is not None
+        assert self.infile is not None
         self.stop_event.set()
         self.thread.join()
         self.infile.close()
 
 
 def prepare_outputs(
     capture_outputs: str, outputs: str, output_prefix: str
-) -> Tuple[Union[TextIO, TailPipe, int], Union[TextIO, TailPipe, int]]:
-    stdout: Union[TextIO, TailPipe, int]
-    stderr: Union[TextIO, TailPipe, int]
+) -> tuple[TextIO | TailPipe | int | None, TextIO | TailPipe | int | None]:
```

Comment:
why tuple instead of typing.Tuple?

Reply:
The `typing` types for builtins (`Dict`, `Tuple`, `List`, etc.) are all deprecated in favor of just using the types themselves.